pkg/virtctl/adm/logverbosity/logverbosity.go

```diff
@@ -0,0 +1,446 @@
+
+	usage += "  # show log-verbosity for virt-handler and virt-launcher\n"
+	usage += "  {{ProgramName}} adm log-verbosity --virt-handler --virt-launcher\n"
+	usage += "  # set log-verbosity for virt-handler to 7 and virt-launcher to 3\n"
+	usage += "  {{ProgramName}} adm log-verbosity --virt-handler=7 --virt-launcher=3\n"
+
+	usage += "  # reset all components to default besides virt-handler which is 7\n"
+	usage += "  {{ProgramName}} adm log-verbosity --reset --virt-handler=7\n"
+	usage += "  # set all components to 3 besides virt-handler which is 7\n"
+	usage += "  {{ProgramName}} adm log-verbosity --all=3 --virt-handler=7\n"
+
+	return usage
+}
+
+// VirtComponent to component name
+// also used by the test file
+func GetComponentNameByVirtComponent(component virtComponent) string {
+	var virtComponentToComponentName = map[virtComponent]string{
+		VirtAPI:        "virt-api",
+		VirtController: "virt-controller",
+		VirtHandler:    "virt-handler",
+		VirtLauncher:   "virt-launcher",
+		VirtOperator:   "virt-operator",
+		All:            "all",
+	}
+	return virtComponentToComponentName[component]
+}
+
+// virtComponent to JSON name
+func getJSONNameByVirtComponent(component virtComponent) string {
+	var virtComponentToJSONName = map[virtComponent]string{
+		VirtAPI:        "virtAPI",
+		VirtController: "virtController",
+		VirtHandler:    "virtHandler",
+		VirtLauncher:   "virtLauncher",
+		VirtOperator:   "virtOperator",
+		All:            "all",
+	}
+	return virtComponentToJSONName[component]
+}
+
+// component name to JSON name
+func getJSONNameByComponentName(componentName string) string {
+	var componentNameToJSONName = map[string]string{
+		"virt-api":        "virtAPI",
+		"virt-controller": "virtController",
+		"virt-handler":    "virtHandler",
+		"virt-launcher":   "virtLauncher",
+		"virt-operator":   "virtOperator",
+		"all":             "all",
+	}
+	return componentNameToJSONName[componentName]
+}
+
+func detectInstallNamespaceAndName(virtClient kubecli.KubevirtClient) (namespace, name string, err error) {
+	// listing KubeVirt CRs in all namespaces and see where it is installed
+	kvs, err := virtClient.KubeVirt("").List(&k8smetav1.ListOptions{})
+	if err != nil {
+		return "", "", fmt.Errorf("could not list KubeVirt CRs in all namespaces: %v", err)
+	}
+	if len(kvs.Items) == 0 {
+		return "", "", errors.New("could not detect a KubeVirt installation")
+	}
+	if len(kvs.Items) > 1 {
+		return "", "", errors.New("invalid kubevirt installation, more than one KubeVirt resource found")
+	}
+	namespace = kvs.Items[0].Namespace
+	name = kvs.Items[0].Name
+	return
+}
+
+func hasVerbosityInKV(kv *v1.KubeVirt) (map[string]uint, error) {
+	verbosityMap := map[string]uint{} // key: component name, value: verbosity
+	// check the logVerbosity field in the KubeVirt CR
+	lvJSON, err := json.Marshal(kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity)
+	if err != nil {
+		return nil, err
```

Comment:
Should be fine to return `verbosityMap` even on error. It reduces the extra check in the `main` function

Reply:
Good point. You are right. It should be fine to return verbosityMap even on error. However, the reason why we need the check below at the `main` is:
```
	if currentLv == nil {
		currentLv = make(map[string]uint)
	}
```

`verbosityMap = nil` in case of no entry of `kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity` like below.
```
	&v1.KubeVirt{
		TypeMeta: k8smetav1.TypeMeta{
			Kind:       "KubeVirt",
			APIVersion: v1.GroupVersion.String(),
		},
		ObjectMeta: k8smetav1.ObjectMeta{
			Namespace: namespace,
			Name:      name,
		},
		Spec: v1.KubeVirtSpec{
			ImageTag: "devel",
			Configuration: v1.KubeVirtConfiguration{
				DeveloperConfiguration: &v1.DeveloperConfiguration{},
			},
		},
	}
```
Now, although this does not make the code simple, I think it is better (more logical) to move the below code to `hasVerbosityInKV` from `main` (add the below code to `hasVerbosityInKV` (right before the return line) and remove it from `main`). And, remain nil at the above return lines, because the error will be handled at the main and the verbosityMap is irrelevant in the case.
```
func hasVerbosityInKV(kv *v1.KubeVirt) (map[string]uint, error) {
	verbosityMap := map[string]uint{} // key: component name, value: verbosity
	// check the logVerbosity field in the KubeVirt CR
	lvJSON, err := json.Marshal(kv.Spec.Configuration.DeveloperConfiguration.LogVerbosity)
	if err != nil {
		return nil, err
	}
	// if there is a logVerbosity field, input the JSON data to the verbosityMap
	if err := json.Unmarshal(lvJSON, &verbosityMap); err != nil {
		return nil, err
	}
	// if map is nil (= logVerbosity field in the KubeVert CR is nil), need initialization to access a key in the map
	if verbosityMap == nil {
		verbosityMap = map[string]uint{}
	}
	return verbosityMap, nil
}
```
What do you think?